src/Pap/ClosureHistoryManager.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace App\Pap;
+
+use App\Entity\Adherent;
+use App\Entity\Pap\Building;
+use App\Entity\Pap\BuildingStatistics;
+use App\Entity\Pap\Campaign;
+use App\Entity\Pap\CampaignStatisticsOwnerInterface;
+use Doctrine\ORM\EntityManagerInterface;
+
+class ClosureHistoryManager
+{
+    private EntityManagerInterface $em;
+
+    public function __construct(EntityManagerInterface $em)
+    {
+        $this->em = $em;
+    }
+
+    public function open(
+        CampaignStatisticsOwnerInterface $objet,
+        Campaign $campaign,
+        \DateTimeInterface $openedAt,
+        Adherent $openedBy
+    ): void {
+        $this->openClose(BuildingStatusEnum::ONGOING, $objet, $campaign, $openedAt, $openedBy);
+    }
+
+    public function close(
+        CampaignStatisticsOwnerInterface $objet,
+        Campaign $campaign,
+        \DateTimeInterface $closedAt,
+        Adherent $closedBy
+    ): void {
+        $this->openClose(BuildingStatusEnum::COMPLETED, $objet, $campaign, $closedAt, $closedBy);
+    }
+
+    public function openClose(
+        string $status,
+        CampaignStatisticsOwnerInterface $objet,
+        Campaign $campaign,
+        \DateTimeInterface $doneAt,
+        Adherent $doneBy
+    ): void {
+        $stats = $objet->getStatisticForCampaign($campaign);
+        if (!$stats) {
```

Comment:
I don't understand why you use a switch case here with only one case and no default.
```suggestion
            if (!$stats && is_a($object, Building::class) ) {
```